### PR TITLE
Replace proposal EXT_multi_draw_arrays with WEBGL_multi_draw and WEBGL_multi_draw_instanced (#2717)

### PR DESCRIPTION
**Pull request by austinEng**
Rename EXT_multi_draw_arrays to ANGLE_multi_draw.
Adds additional entry points for Instanced and InstancedBaseVertex Multi* draw call variants.
Add additional builtins gl_DrawID, gl_InstanceID, and gl_BaseVertex.

### CODE REVIEW

File: extensions/proposals/ANGLE_multi_draw/extension.xml [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+
+      <p>CAD vendors rendering large models comprised of many individual parts face scalability issues issuing large numbers of draw calls from WebGL. This extension reduces draw call overhead by allowing better batching.</p>
+    </div>
+
+    <features>
+      <feature>The <code>*Offset</code> arguments to the new entry points choose the starting offset into their respective typed arrays or sequences. This primarily avoids allocation of temporary typed array views.</feature>
+
+      <feature>The <code>multiDrawArraysANGLE</code> and <code>multiDrawElementsANGLE</code> entry points are added. These provide a counterpoint to instanced rendering and are more flexible for certain scenarios. They behave as to multiple calls to <code>drawArrays</code> and <code>drawElements</code> except they handle multiple lists of arguments in one call.</feature>
+
+      <feature>The <code>gl_DrawID</code> builtin is added to the shading language. For any Multi* draw call variant, the index of the draw <code>i</code> may be read by the vertex shader as <code>gl_DrawID</code>.</feature>
```
[the remark this answers is not in the document]
Comment (austinEng) [reply extensions/proposals/ANGLE_multi_draw/extension.xml]: Done. It should be 0. Added a test in `runShaderTests()`

File: extensions/proposals/ANGLE_multi_draw/extension.xml [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+
+      <p>This extension additionally adds the <code>gl_DrawID</code> builtin for the new Multi* draw commands</p>
+
+      <p>CAD vendors rendering large models comprised of many individual parts face scalability issues issuing large numbers of draw calls from WebGL. This extension reduces draw call overhead by allowing better batching.</p>
+    </div>
+
+    <features>
+      <feature>The <code>*Offset</code> arguments to the new entry points choose the starting offset into their respective typed arrays or sequences. This primarily avoids allocation of temporary typed array views.</feature>
+
+      <feature>The <code>multiDrawArraysANGLE</code> and <code>multiDrawElementsANGLE</code> entry points are added. These provide a counterpoint to instanced rendering and are more flexible for certain scenarios. They behave as to multiple calls to <code>drawArrays</code> and <code>drawElements</code> except they handle multiple lists of arguments in one call.</feature>
```
[the remark this answers is not in the document]
Comment (austinEng) [reply extensions/proposals/ANGLE_multi_draw/extension.xml]: Done

File: sdk/tests/conformance/extensions/angle-multi-draw.html [hunk 3]
```diff
@@ -0,0 +1,831 @@
...
+  function getContextVersion() {
+    if (typeof WebGLRenderingContext !== "undefined" && gl instanceof WebGLRenderingContext) {
+      return 1;
+    } else if (typeof WebGL2RenderingContext !== "undefined" &&  gl instanceof WebGL2RenderingContext) {
+      return 2;
+    } else {
+      testFailed('Unsupported rendering context');
+      throw 'Unsupported rendering context';
+    }
+  }
```
[the remark this answers is not in the document]
Comment (kainino0x) [reply sdk/tests/conformance/extensions/angle-multi-draw.html]: getDefault3DContextVersion seems good.

File: extensions/proposals/ANGLE_multi_draw/extension.xml [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+    </features>
+  </overview>
+
+  <idl xml:space="preserve">
+
+interface mixin ANGLE_multi_draw_mixin {
+  void multiDrawArraysANGLE(
+      GLenum mode,
+      (Int32Array or sequence&lt;GLint&gt;) firstsList, GLuint firstsOffset,
+      (Int32Array or sequence&lt;GLsizei&gt;) countsList, GLuint countsOffset,
```
[the remark this answers is not in the document]
Comment (kainino0x) [reply extensions/proposals/ANGLE_multi_draw/extension.xml]: You're right.

**Comment by austinEng** [comment]
@jdashg These changes remove the BaseVertex part.

**Comment by jdashg** [comment]
I will check tomorrow.

On Nov 7, 2018 8:18 PM, "Ken Russell" <notifications@github.com> wrote:

*@kenrussell* approved this pull request.

Nice work @austinEng <https://github.com/austinEng> . @jdashg
<https://github.com/jdashg> are your comments resolved with the removal of
the BaseVertex variants?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/KhronosGroup/WebGL/pull/2735#pullrequestreview-172801285>,
or mute the thread

File: extensions/proposals/ANGLE_multi_draw/extension.xml [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+      GLenum mode,
+      (Int32Array or sequence&lt;GLint&gt;) countsList, GLuint countsOffset,
+      GLenum type,
+      (Int32Array or sequence&lt;GLsizei&gt;) offsetsList, GLuint offsetsOffset,
+      (Int32Array or sequence&lt;GLsizei&gt;) instanceCountsList, GLuint instanceCountsOffset,
+      GLsizei drawcount);
+};
+
+[NoInterfaceObject]
+ANGLE_multi_draw_1
```
Comment (jdashg) [inline extensions/proposals/ANGLE_multi_draw/extension.xml]: ANGLE_multi_draw

File: extensions/proposals/ANGLE_multi_draw/extension.xml [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+{
+};
+
+[NoInterfaceObject]
+ANGLE_multi_draw_2
```
Comment (jdashg) [inline extensions/proposals/ANGLE_multi_draw/extension.xml]: ANGLE_multi_draw_instanced, but in its own extension file.

File: extensions/proposals/ANGLE_multi_draw/extension.xml [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+{
+};
+
+ANGLE_multi_draw_1 includes ANGLE_multi_draw_mixin;
+ANGLE_multi_draw_2 includes ANGLE_multi_draw_mixin;
```
Comment (jdashg) [inline extensions/proposals/ANGLE_multi_draw/extension.xml]: Don't include multi_draw in multi_draw_instanced.

**Comment by austinEng** [comment]
@jdashg This has now been split into two extensions `WEBGL_multi_draw` and `WEBGL_multi_draw_instanced`.

**Comment by jdashg** [comment]
Thanks!

**Pull request merged**